### PR TITLE
Add generics feature at Stark Proof System

### PR DESCRIPTION
**Pull request by cyl19970726**

### CODE REVIEW

File: starky/src/f3g.rs [hunk 1]
```diff
@@ -30,15 +34,40 @@ impl F3G {
             dim: 3,
         }
     }
+}
+
+impl FnG for F3G {
```
Comment (eigmax) [inline starky/src/f3g.rs:39]: Better to name FnG as FieldExtension.

File: starky/src/constant.rs [hunk 2]
```diff
@@ -27,16 +33,17 @@ lazy_static::lazy_static! {
         m
     };
 
-    pub static ref SHIFT: F3G = F3G::from(FGL::from(49u64));
-    pub static ref SHIFT_INV: F3G = F3G::inv(SHIFT.clone());
-    pub static ref MG: (Vec<F3G>, Vec<F3G>) = {
-        let nqr = F3G::from(FGL::from(7u64));
-        let rem = 2usize.pow(32) - 1;
+    pub static ref SHIFT: Fr = Fr(FrRepr::from(49u64));
```
Comment (ibmp33) [inline starky/src/constant.rs:36]: pub static ref SHIFT: Fr = Fr::from(49u64);

File: starky/src/f3g.rs [hunk 3]
```diff
@@ -137,7 +166,22 @@ impl F3G {
...
 
     #[inline]
-    pub fn batch_inverse(elems: &[Self]) -> Vec<Self> {
+    fn inv(self) -> Self {
+        self._inv()
+    }
+
+    #[inline]
+    fn as_int(&self) -> u64 {
+        self._as_int()
```
Comment (eigmax) [inline starky/src/f3g.rs:175]: Can you move the function body of _xxxx(e.g. _as_int) into xxxx, and remove the _xxxx, hence the code would be more simplication

File: starky/src/merklehash.rs [hunk 4]
```diff
@@ -154,7 +154,7 @@ impl MerkleTreeGL {
 impl MerkleTree for MerkleTreeGL {
     type BaseField = FGL;
     type MTNode = ElementDigest<4>;
-
+    type FnG = F3G;
```
Comment (eigmax) [inline starky/src/merklehash.rs:157]: Maybe we can name this type alias as ExtendField.

**Comment by eigmax** [comment]
Thanks！

**Pull request merged**